### PR TITLE
Make completion more resilient to bad locs

### PR DESCRIPTION
**Pull request by jez**
### Motivation

The root cause of some crashes I was seeing  didn't turn out to be my fault,
but it was work I did thinking that it was my fault, and I think we're better
off having these checks, so I figured I'd pull them into their own change and
land them.

### Test plan

No automated tests, because I can't actually imagine a case where this happens
normally.

### CODE REVIEW

File: main/lsp/LSPConfiguration.cc [hunk 1]
```diff
@@ -99,7 +99,14 @@ core::Loc LSPConfiguration::lspPos2Loc(const core::FileRef fref, const Position
     core::Loc::Detail reqPos;
     reqPos.line = pos.line + 1;
     reqPos.column = pos.character + 1;
-    auto offset = core::Loc::pos2Offset(fref.data(gs), reqPos);
```
Comment (DarkDimius) [inline main/lsp/LSPConfiguration.cc]: does this ~reimplement pos2Offset?
could we make it return `optional` instead?

**Comment by DarkDimius** [comment]
Did you find a way to repro? would be nice to add a test, but happy to not block on it

**Comment by jez** [comment]
> Did you find a way to repro? would be nice to add a test, but happy to not block on it

The original regression was unrelated to my change, and jvilk fixed it:

https://github.com/sorbet/sorbet/pull/1975

I think the only way this that completion could *possibly* crash here is if we made completion requests delayable, and so the completion cursor position gets delayed to operate on an edited version of that file where the position doesn't make sense anymore.

We decided to not go ahead with delayable completion requests, because we didn't have to. But even still, I figured it couldn't hurt to be cautious here. So the only way we could write a test for this would be to manually record a misbehaving LSP `.rec` test, which seems like overkill.

File: main/lsp/LSPConfiguration.cc (hunk 1 above)
Comment (jez) [reply main/lsp/LSPConfiguration.cc]: Good suggestion! In the process of refactoring i uncovered an ENFORCE failure that only manifests for files with a class def on the last line of a file with no line break at the end of the file. We don't crash on that anymore!

**Pull request merged**